cmd/dep/gvt_importer.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/golang/dep"
+	"github.com/golang/dep/internal/gps"
+	"github.com/pkg/errors"
+)
+
+const gvtPath = "vendor" + string(os.PathSeparator) + "manifest"
+
+type gvtImporter struct {
+	*baseImporter
+	gvtConfig gvtManifest
+}
+
+func newGvtImporter(logger *log.Logger, verbose bool, sm gps.SourceManager) *gvtImporter {
+	return &gvtImporter{baseImporter: newBaseImporter(logger, verbose, sm)}
+}
+
+type gvtManifest struct {
+	Deps []gvtPkg `json:"dependencies"`
+}
+
+type gvtPkg struct {
+	ImportPath string
+	Repository string
+	Revision   string
+	Branch     string
+}
+
+func (g *gvtImporter) Name() string {
+	return "gvt"
+}
+
+func (g *gvtImporter) HasDepMetadata(dir string) bool {
+	y := filepath.Join(dir, gvtPath)
+	if _, err := os.Stat(y); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (g *gvtImporter) Import(dir string, pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	err := g.load(dir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return g.convert(pr)
+}
+
+func (g *gvtImporter) load(projectDir string) error {
+	g.logger.Println("Detected gvt configuration files...")
+	j := filepath.Join(projectDir, gvtPath)
+	if g.verbose {
+		g.logger.Printf("  Loading %s", j)
+	}
+	jb, err := ioutil.ReadFile(j)
+	if err != nil {
+		return errors.Wrapf(err, "unable to read %s", j)
+	}
+	err = json.Unmarshal(jb, &g.gvtConfig)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse %s", j)
+	}
+
+	return nil
+}
+
+func (g *gvtImporter) convert(pr gps.ProjectRoot) (*dep.Manifest, *dep.Lock, error) {
+	g.logger.Println("Converting from vendor/manifest ...")
+
+	packages := make([]importedPackage, 0, len(g.gvtConfig.Deps))
+	for _, pkg := range g.gvtConfig.Deps {
+		// Validate
+		if pkg.ImportPath == "" {
+			err := errors.New("invalid gvt configuration, ImportPath is required")
+			return nil, nil, err
+		}
+
+		if pkg.Revision == "" {
+			err := errors.New("invalid gvt configuration, Revision is required")
+			return nil, nil, err
+		}
+
+		var contstraintHint = ""
+		if pkg.Branch != "master" {
```

Comment:
I added this line because it actually affected the behavior. 
for example, I ran it on a `gvt` project with this in the `manifest`:
```json
{                                                                        
  "importpath": "github.com/blang/semver",                                                                                                                 
  "repository": "https://github.com/blang/semver",                     
  "vcs": "",                                                           
  "revision": "aea32c919a18e5ef4537bbd283ff29594b1b0165",              
  "branch": "master"                                                   
}, 
```
by passing the `"master" ` as a `ConstraintHint`, `dep init`  chose a newer version:
```toml
[[projects]]
  name = "github.com/blang/semver"
  packages = ["."]
  revision = "2ee87856327ba09384cabd113bc6b5d174e9ec0f"
  version = "v3.5.1"
```

but by not passing `"master"` as a hint with the check above, `dep init` chose the correct version:
```toml
[[projects]]
  name = "github.com/blang/semver"
  packages = ["."]
  revision = "aea32c919a18e5ef4537bbd283ff29594b1b0165"
  version = "v3.1.0"
````